### ISSUE TITLE
【Mosaic bug】 incorrect image segmentation affects training

### ISSUE THREAD

**Issue by sizhongyibanhts** [ultralytics/ultralytics#21927]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Ultralytics YOLO Component
No response

Bug

### Problem Phenomenon（问题现象）

Mosaic is a highly effective training method, but I've noticed some problems with it.（mosaic是非常有效的训练手段，但我发现了一些问题）。You can visit this page to see a more complete analysis of the [problem.](https://blog.csdn.net/weixin_37763484/article/details/151152493)

I encountered an issue with mosaic augmentation, where it incorrectly splits data and generates illogical bounding boxes. Today, while training an insulator recognition model, I found this problem to be severe—almost every ten images contains a serious error. In these erroneous cases, the bounding boxes do not contain any insulators at all, but instead only include irrelevant background regions or other electrical equipment.（我遇到了一个关于mosaic增强的问题，mosaic会错误的切分数据，产生不符合逻辑的标注框。今天我在训练绝缘子识别模型，发现这个现象很严重，几乎每十张图片就会出现一个严重的错误。这种错误的标注框中，根本不包含绝缘子，只包含一些无关的背景或其他电力设备。）

The original image is as follows（原始图像如下）：

<img width="751" height="674" alt="Image" src="https://github.com/user-attachments/assets/db88bb11-b8fb-4fd6-b90d-74ac98a0a03d" />

After applying mosaic data augmentation（使用mosaic数据增强后）：

<img width="591" height="527" alt="Image" src="https://github.com/user-attachments/assets/6d59af94-c3ee-4915-a2c5-a35291cd3de4" />

### Problem Analysis 问题分析

If you look closely at the insulator in the bottom right corner of the augmented image, the position of the bounding box is clearly offset（你可以仔细观察，增强之后图像右下角的绝缘子，标注框位置明显偏移）。

This would likely have a severe impact on the model's training. There are even more severe cases where the insulator is no longer visible within the augmented bounding box at all. This happens because my target object is located along the diagonal of its bounding box. When the mosaic augmentation process truncates the box, it's highly likely to crop a diagonal region that doesn't actually contain the target（这应该会严重影响到模型训练，还有一些更严重的情况，增强后的标注框中已经无法看到绝缘子。这是因为我的目标位于标注框的对角线上，当你的mosaic增强把坐标框截断的时候，非常有可能截取到不包含目标的对角线区域。）

### Solution Recommendations（解决建议）

When I turned off mosaic, similar issues became much rarer (almost non-existent), so mosaic must be too aggressive.（（当关闭马赛克后，类似的情况就少了很多（几乎没有），因此应该是马赛克过于粗暴了。）

Since my targets are diagonal bounding boxes, mosaic cropping needs to recalculate the new coordinates based on the intersection of the diagonal and the cropping boundary.（由于我的目标是斜跨标注框对角线的，因此在马赛克截取的时候，需要重新根据对角线位置计算新的坐标框（计算截取位置与对角线的交点）。）

Alternatively, I should either not use mosaic for my scenario, or ensure that mosaic augmentation completely preserves the bounding boxes for certain special classes.（或者在我的场景下，不要使用mosaic增强，或者保证马赛克增强的时候，把某些特殊类别的坐标框完整的保留。）

Another option is to use mosaic less or turn it off early. However, the model's confusion and wasted training that happens while it's on would be hard to correct.（或者少用一些马赛克，提前关闭马赛克，但是未关闭的时候，造成的模型误会和训练浪费，恐怕难以纠正。）

Or, when mosaic crops more than 50% of a target's area, automatically black out the remaining part.（或者当马赛克将某个目标截取了超过50%面积的时候，自动将目标剩余的部分涂黑。）

### Environment

windows11

This bug is not related to the computer system.

这个bug与计算机系统无关

### Minimal Reproducible Example

你可以在这个网页中，看到问题的截图  [link](https://blog.csdn.net/weixin_37763484/article/details/151152493)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sizhongyibanhts, thank you for your interest in Ultralytics 🚀! This is an automated response to help triage your report quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

For your mosaic-related report, could you please share a minimal reproducible example including:
- exact ultralytics version (`yolo -v` or `python -c "import ultralytics, torch; print(ultralytics.__version__, torch.__version__)"`)
- a small sample (1–3) of original images and their labels exhibiting the issue
- the exact training/augment settings used (mosaic on/off schedule, mosaic scale/probability, any custom hyp.yaml)
- the command or Python snippet you ran and full logs
- a screenshot or saved batch showing the incorrect boxes after augmentation (`plots=True` or `save=True` if applicable)
- OS, CUDA, GPU model, and PyTorch build details

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed report and screenshots 🙏. Once you share the MRE and version details, we’ll be able to reproduce and investigate the mosaic behavior more quickly.

**Comment by Y-T-G** [comment]
This isn't a bug. It's a consequence of bounding boxes not having precise contour information. There's no easy way to tell from just a box what the exact contour of your actual object is to recalculate the new box. Bounding boxes don't have that information.

If you want better fitting boxes, you can use  instance segmentation labels to train a detection model. This allows recalculating new box from the mask.

**Comment by sizhongyibanhts** [comment]
You mentioned "using the labels from a segmentation model to train a detection model," which I haven't tried before. May I ask if it's possible to directly train a segmentation model instead?

**Comment by Y-T-G** [comment]
You can train both detection and segmentation models using a segmentation dataset.

**Comment by sizhongyibanhts** [comment]
Thanks for the thorough analysis—this is a known side-effect of Mosaic with axis-aligned boxes: after the affine step we clip boxes and filter by area ratio (see the RandomPerspective box filter, area_thr=0.1) which can leave little or no foreground when objects run along a diagonal; details in the [Mosaic augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/#mosaic-mosaic) and the [transform/box-candidate reference](https://docs.ultralytics.com/reference/data/augment/#box_candidates-staticmethod). To mitigate, reduce Mosaic probability and disable it early, or switch to OBB so boxes rotate with objects (see the [OBB task docs](https://docs.ultralytics.com/tasks/obb/)); 

for example:

yolo train model=yolo11n.pt data=insulator.yaml epochs=100 mosaic=0.2 close_mosaic=50 mixup=0.1 cutmix=0.1

If you still suspect a bug, please attach a tiny reproducible sample (4–8 images with labels) plus your exact train command and yolo checks output (text, not screenshots), and we’ll investigate or you can open a new issue for visibility.

**Comment by sizhongyibanhts** [comment]
Thank you for your explanation. I have a question: when training an object detection model using a segmentation dataset, could the issue mentioned earlier still occur?

My objects are located diagonally within the bounding boxes. As I understand it, even when using a segmentation dataset, during training of the detection model with Mosaic augmentation enabled, the problem shown in the previous image could still happen, right?

**Comment by glenn-jocher** [comment]
Short answer: mostly no—when you train a detection model on a segmentation dataset, the pipeline applies Mosaic/affine to the masks and then re-derives the boxes from the transformed masks via `apply_segments`, which avoids background-only boxes; see [apply_segments in RandomPerspective](https://docs.ultralytics.com/reference/data/augment/#apply_segments) and the [RandomPerspective transform flow](https://docs.ultralytics.com/reference/data/augment/#call) where boxes are recomputed from masks and a looser `area_thr` is used; you’ll still get axis-aligned boxes, so diagonally oriented objects can look “loose”—use OBB or train segmentation if you need rotation fidelity.

**Comment by sizhongyibanhts** [comment]
**This is an official response that provides a more comprehensive answer. However, it was posted in another issue that has already been closed, so I've copied it here for everyone's reference.**
**这是官方回复，给出了更全面的答案，但是这个回复在另一个已经关掉的issue中，我复制过来，供大家参考**

Good questions. **OBB doesn’t inherently reduce accuracy**; for diagonally oriented targets it often improves localization when you train and evaluate with OBB labels and metrics—see this brief [OBB overview](https://www.ultralytics.com/blog/what-is-oriented-bounding-box-obb-detection-a-quick-guide). For a quick test:
（很好的问题。OBB 本身并不会降低精度；对于对角线方向的目标，在使用 OBB 标签和指标进行训练和评估时，通常还能提升定位精度——详见这篇简要的 OBB 概述。如需快速测试：）

yolo obb train model=yolo11n-obb.pt data=insulator-obb.yaml epochs=100 imgsz=640

**If you prefer AABB, a minimal safeguard** is to drop heavily truncated boxes after the affine step by raising the area filter in ultralytics/data/augment.py inside RandomPerspective.__call__:
（如果你更倾向于使用 AABB，一个最基本的防护措施是在仿射变换后剔除被严重截断的框，方法是在 ultralytics/data/augment.py 的 RandomPerspective.call 函数中提高面积过滤阈值：）

# replace
i = self.box_candidates(box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.10)
# with, e.g.
i = self.box_candidates(box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.30)

**You can also reduce geometric motion to limit cuts:**
（你也可以减少几何变换幅度来限制裁剪：）

yolo train model=yolo11n.pt data=insulator.yaml epochs=100 translate=0.02

If you need format conversions for OBB↔AABB during evaluation or tooling, the box utilities are referenced here: [OBB box conversion ops](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.xywhr2xyxyxyxy).
如果在评估或工具处理中需要 OBB 与 AABB 之间的格式转换，相关边界框操作工具可参考此处：OBB box conversion ops。

**Comment by glenn-jocher** [comment]
Thanks for sharing this summary. Two quick additions: to audit what the model actually trains on, open the saved batch previews at `runs/.../train_batch*.jpg`; and if you switch label types (AABB → segmentation or OBB), delete any `*.cache` files in your `labels` folders to avoid stale annotations. If you move to OBB, ensure labels follow the YOLO OBB format described in the [OBB datasets overview](https://docs.ultralytics.com/datasets/obb/). Since this thread is closed, please open a new issue with a 4–8 image repro plus your exact command and `yolo checks` output if you need follow‑up.
